numpy/lib/index_tricks.py

```diff
@@ -877,7 +883,15 @@ def fill_diagonal(a, val, wrap=False):
         step = 1 + (cumprod(a.shape[:-1])).sum()
 
     # Write the value out into the diagonal.
-    a.flat[:end:step] = val
+    # if it's a 2D square array with wrap, manually fill in diagonals so
+    # rows aren't skipped between diagonals
+    flag = a.ndim == 2 and a.shape[0] == a.shape[1] and wrap
+    if flag:
+        for i in range(a.shape[0]):
+            a[i][(i+offset)%a.shape[0]] = val
```

Comment:
Thanks for the PR!

Just to understand: Why the special case for 2d square arrays here? For non-square arrays we already "skip" a row before wrapping, so it seems like we should also skip it for square arrays with an offset, to be consistent. I suggest always just doing `a.flat[offset:end:step] = val`, no special case.

(I'm not sure "why" we skip a row, except that that is how the indexing trick used here works when extending the algorithm from square to non-square arrays. I suspect no-one really uses "wrap". It was added in https://github.com/numpy/numpy/pull/306, apparently as a way to *prevent* wrapping rather than allow it.).